### ISSUE TITLE
feature request: z-x-y coordinate urls

### ISSUE THREAD

**Issue by autur** [lovasoa/dezoomify-rs#6]
In case this isn't already an option, are there any plans to support urls with the z-x-y coordinates, as is currently possible with the web version of Dezoomify?

For example, this works for the web version but not rs: http://0.tiles.dx.artsmia.org/1743/7/{{X}}/{{Y}}.jpg
BTW, the .json url does not work for either, so I had to use the z-x-y url: https://tiles.dx.artsmia.org/1743.json

If this is going to be implemented at some point, I hope you can do an automatic fallback for the z value (zoom level), e.g., if 7 is NA, then try 6, then 5, etc. The custom .yaml option is nice, but if a url can be provided, that makes it much less cumbersome, especially for multiple images.

It would be super nice to be able to do everything from the command line with dezoomify-rs, as opposed to the browser. I hope we are gradually moving in that direction :)

**Update:** I used this example because the download button on the site gave the 800px version, but a quick edit of the download url from "800" to "full" gives the full size (https://0.api.artsmia.org/full/1743.jpg). Regardless of this easy fix, it would be nice to have the feature I mentioned for other cases with no easy fixes like this.

**Comment by lovasoa** [comment]
What I would like to do is to simplify dramatically the tiles.yaml format, so that you could write something like 

```yaml
url_template: "http://0.tiles.dx.artsmia.org/1743/7/{{X}}/{{Y}}.jpg"
width: 6824
height: 17025
tile_size: 256
```

**Comment by lovasoa** [comment]
I started thinking about what we would need to implement url-template-only dezooming in an efficient manner.

Here is a state machine we could implement : 
[<img width="1326" alt="image" src="https://user-images.githubusercontent.com/552629/63446061-d22f3580-c439-11e9-8e02-4ae66ca9d5c6.png">](https://wbo.openode.io/boards/gdwQn5gCdqq0LbSk36myO3-1nXSXTJK-E1TdgdchtbY-#27,0,0.7)

Now we have to find the best way to implement this in a generic manner; that is, in a manner that avoids code duplication between the generic dezoomer and the others.

**Comment by autur** [comment]
Ah, interesting, so this is the flow process for Dezoomify. But I thought this was already implemented in the browser version, no?

**Comment by lovasoa** [comment]
This process is required only when the tile size and the image size is not known in advance.
In the current version of dezoomify-rs (and dezoomify), this information is always available before starting tile downloads, and a more efficient parallel workflow can be adopted (more work can happen in parallel).
What we are seeking is a generic implementation that would implement both workflows, without duplicating the logic that is common between them.

The browser version of dezoomify uses a dichotomy on tile loads to first determine the tile size and image size, and then uses the parallel workflow. This is however not optimal, as it makes more requests than necessary.

**Comment by lovasoa** [comment]
I implemented a generic dezoomer in dezoomify-rs v1.4.0. Usage instructions are in the README.

**Comment by autur** [comment]
Hey, coming back to report some flaws with the coordinate URL option: 

I've had many cases where the final image has incomplete tiles and I suspect that is because there are other variables in the URL that change aside from X and Y. 

Some examples (link + zoomify url):

1. http://www.clevelandart.org/art/1924.231
https://piction.clevelandart.org/zoomify/948D83A9913C3E229ECCF64817A48A5D/TileGroup0/5-{{X}}-{{Y}}.jpg

2. http://www.clevelandart.org/art/1998.168#
https://piction.clevelandart.org/zoomify/467229821B505A1D84255F24F529D95B/TileGroup1/5-{{X}}-{{Y}}.jpg  (This one can't get any tile)

3. https://www.ngv.vic.gov.au/explore/collection/work/3161/
https://content.ngv.vic.gov.au/col-images/zooms/Dd103476/TileGroup0/5-{{X}}-{{Y}}.jpg

In the last example, if you zoom in and browse the image, you will find tiles where TileGroup varies from 1 to 0, and the next value before X-Y varies from 4 to 5, as shown:
https://content.ngv.vic.gov.au/col-images/zooms/Dd103476/TileGroup1/5-3-11.jpg
https://content.ngv.vic.gov.au/col-images/zooms/Dd103476/TileGroup0/4-1-0.jpg

Are you aware of this? 

The other thing I hope to see implemented is a fall-back if the highest zoom is NA, like if the URL with 5-X-Y fails, then try 1 below (4-X-Y), etc. This would make it much more convenient :)

**Comment by lovasoa** [comment]
Hello! All of your URLs are using the zoomify protocol, that is more complex than just an x and a y in the URL. As you noticed yourself, the TlieGroup number has to be set to the right value for each tile. Fortunately, dezoomify-rs DOES SUPPORT the zoomify protocol. You have to give it the URL of the ImageProperties.xml file. Refer to this project's README, and if it's not clear, you can open a new issue and I will improve the usage instructions.

**Comment by autur** [comment]
My bad, I was not aware of the zoomify xml option, which works nicely here. Should have read the new documentation. Sorry
